app/assets/javascripts/d3/stacked_bar.coffee

```diff
@@ -173,6 +152,42 @@ D3.stacked_bar =
         result.unshift(1990)
       result
 
+    display_legend: =>
+      $(@container_selector()).find('.legend').remove()
+
+      series_for_legend = @prepare_legend_items()
+      legend_columns = if series_for_legend.length > 6 then 2 else 1
+
+      @draw_legend
+        svg: @svg
+        series: series_for_legend
+        width: @width
+        vertical_offset: @series_height + @legend_margin
+        columns: legend_columns
+
+    prepare_legend_items: =>
+      # Prepare legend
+      # remove duplicate target series. Required for backwards compatibility.
+      # When we'll drop the old charts we should use a single serie as target
+      # rather than two.
+      #
+      # Also checks if series have a significant value i.e. a value larger than
+      # 0.0000001 otherwise it doesn't get added to the legend.
+      target_lines = []
+      series_for_legend = []
+
+      for s in @series
+        label = s.get 'label'
+        if s.get 'is_target_line'
+          if _.indexOf(target_lines, label) == -1
+            target_lines.push label
+            series_for_legend.push s
+          # otherwise the target line has already been added
+        else if (Math.round(s.future_value() * 10e6) / 10e6) > 0
```

Comment:
> Should this also check if the present value is greater than zero? It could be that the future value is zero, but the present is non-zero?

I don't think that's really needed, is it? In the start of your scenario your present == future. So when changing sliders, the future changes and therefor your bar chart. However, if you start to exclude certain parts of the energy flow from the present (is this possible?) than you wouldn't see them as legend items. I'll look into it 👍

Reply:
> I don't think that's really needed, is it?

I managed to find an example:

1. Open "Sources of electricity in households" chart
2. Set "Households" → "Solar panels" → "PV panels" to `0.0`

<img width="620" alt="screen shot 2017-09-27 at 10 07 37" src="https://user-images.githubusercontent.com/4383/30905151-e20d73ae-a36b-11e7-845e-9e747c86cc9f.png">

The present has a non-zero "Solar panels" value (the green part of the present bar), but the legend item has disappeared because the future value is zero.

Reply:
![screen shot 2017-09-27 at 13 16 11](https://user-images.githubusercontent.com/2676542/30910683-100c950e-a386-11e7-89d9-ec3d69d7a30b.png)

I managed to fix it with the recent changes 👍